compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirContractChecker.kt

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.contracts.FirResolvedContractDescription
+import org.jetbrains.kotlin.fir.declarations.FirContractDescriptionOwner
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.expressions.*
+
+object FirContractChecker : FirFunctionChecker() {
+    private const val UNEXPECTED_CONSTRUCTION = "unexpected construction in contract description"
+
+    override fun check(declaration: FirFunction<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration !is FirContractDescriptionOwner ||
+            declaration.contractDescription !is FirResolvedContractDescription
+        ) {
+            return
+        }
+
+        // Any statements that [ConeEffectExtractor] cannot extract effects will be in `unresolvedEffects`.
+        for (statement in (declaration.contractDescription as FirResolvedContractDescription).unresolvedEffects) {
```

Comment:
Should any unresolved effects be considered error?

Reply:
That is what I was wondering as well... AFAIK contracts are all compile-time, so there really shouldn't be anything in there except an effect extracted by `ConeEffectExtractor`. Any other statement should be invalid.

What would happen if you just reported an error on every statement in `unresolvedEffects`?

Reply:
I already tried that as well, but it raised too many false alarms. Apparently, legal expressions for `contract` that has yet not extracted properly would be in `unresolvedEffects` as well. So we need one more filtering if it's there because it's illegal construction or it's there because we didn't extract effects (yet).

Reply:
cc @demiurg906

Reply:
Can you please provide examples where legal statements are marked as unresolved effects?

Reply:
It turns out that the locations, i.e., sources of reports were wrong, so I thought they were regressions. :\

Reply:
Regarding report locations, sometimes it's just `contract`, or sometimes it's where the original error, e.g., unresolved reference, happens. Anyway, the best place to nail down the source is when `ConeEffectExtractor` returns `null` for effects. Each case should have different error messages as well. I think it will require more investigations/comparisons. Will do so in a follow-up.